File: fixtures/javascript/controller-service-provider/src/provider/stripeProvider.js
// Concrete provider: CommonJS require bindings and module.exports.

const { PaymentProvider } = require("./provider.js");
const { recordEvent: auditEvent } = require("../shared.js");

class StripeProvider extends PaymentProvider {
    label = "stripe";
    apiKey = "";

    constructor(apiKey) {
        super();
        this.apiKey = apiKey;
    }

    refund(amountCents) {
        this.charge(amountCents);
        auditEvent("refund");
    }

    charge(amountCents) {
        if (amountCents <= 0) {
            throw new Error("amount must be positive");
        }
    }
}

module.exports = { StripeProvider };

File: fixtures/python/controller-service-provider/src/provider/stripe_provider.py
from provider.provider import PaymentProvider
from shared import record_event as audit_event


class StripeProvider(PaymentProvider):
    label = "stripe"

    def __init__(self, api_key):
        self.api_key = api_key

    def refund(self, amount_cents):
        self.charge(amount_cents)
        audit_event("refund")

    def charge(self, amount_cents):
        if amount_cents <= 0:
            raise ValueError("amount must be positive")
